cmd/mt-control-server/handlers.go

```diff
@@ -0,0 +1,339 @@
+	for _, expr := range filter.Expressions {
+		filtersByTag[expr.GetKey()] = append(filtersByTag[expr.GetKey()], expr)
+	}
+
+	numPartitions := request.NumPartitions
+	if numPartitions < 1 {
+		numPartitions = producer.numPartitions()
+	}
+
+	scanPartition := func(partition int32) {
+		q := fmt.Sprintf("SELECT id, orgid, name, interval, unit, mtype, tags, lastupdate FROM %s WHERE partition = ?", cass.Config.ArchiveTable)
+
+		var id, name, unit, mtype string
+		var orgId, interval int
+		var lastupdate int64
+		var tags []string
+
+		var defs []schema.MetricDefinition
+
+		maxLastUpdate := time.Now().Unix()
+
+		session := cass.Session.CurrentSession()
+		iter := session.Query(q, partition).Iter()
+
+	ITER:
+		for iter.Scan(&id, &orgId, &name, &interval, &unit, &mtype, &tags, &lastupdate) {
+
+			// Quick filter on lastupdate
+			if filter.From > 0 && lastupdate < filter.From {
+				continue ITER
+			}
+
+			if filter.To > 0 && lastupdate > filter.To {
+				continue ITER
+			}
+
+			// Tags to map
+			tagMap := make(map[string]string)
+			for _, tag := range tags {
+				tagParts := strings.SplitN(tag, "=", 2)
+				if len(tagParts) != 2 {
+					continue
+				}
+				tagMap[tagParts[0]] = tagParts[1]
+			}
+
+			idTagLookup := func(_ schema.MKey, tag, value string) bool {
+				return tagMap[tag] == value
+			}
+
+			mkey, err := schema.MKeyFromString(id)
+			if err != nil {
+				log.Errorf("job=%s: load() could not parse ID %q: %s -> skipping", logId, id, err)
+				continue
+			}
+
+			for _, expr := range filter.Expressions {
+				matchFunc := expr.GetMetricDefinitionFilter(idTagLookup)
+				if matchFunc(mkey, name, tags) == tagquery.Fail {
+					continue ITER
+				}
+			}
+
+			if orgId < 0 {
+				orgId = int(idx.OrgIdPublic)
+			}
+
+			def := schema.MetricDefinition{
+				Id:         mkey,
+				OrgId:      uint32(orgId),
+				Partition:  partition,
+				Name:       name,
+				Interval:   interval,
+				Unit:       unit,
+				Mtype:      mtype,
+				Tags:       tags,
+				LastUpdate: util.MinInt64(maxLastUpdate, lastupdate+request.LastUpdateOffset),
+			}
+
+			defs = append(defs, def)
+
+		}
+		if err := iter.Close(); err != nil {
+			log.Errorf("job=%s: could not close iterator for partition = %d: %s", logId, partition, err.Error())
+		}
+
+		log.Infof("job=%s: Restoring %d series from partition %d", logId, len(defs), partition)
+
+		if len(defs) == 0 {
+			return
+		}
+
+		// Chunk into batches
+		var defBatches [][]schema.MetricDefinition
+		for i := 0; i < len(defs); i += maxDefsPerMsg {
+			end := i + maxDefsPerMsg
+
+			if end > len(defs) {
+				end = len(defs)
+			}
+
+			defBatches = append(defBatches, defs[i:end])
+		}
+
+		for _, batch := range defBatches {
+			cm := &schema.ControlMsg{
+				Defs: batch,
+				Op:   schema.OpRestore,
+			}
+
+			msg, err := makeMessage(partition, schema.OpRestore, cm)
+			if err != nil {
+				log.Warnf("job=%s: Failed to encode ControlMsg from partition = %d, err = %s", logId, partition, err)
```

Comment:
why are these warnings? I think we should be consistent: any time we cannot (fully) satisfy the user's request, we should report an error, not a warning. this is what we seem to do elsewhere too.

Reply:
This is an asynchronous operation (as it can take a very long time). A job id is returned to the caller to check the logs (if they wish). Are you just suggesting that this be logged at error?

Reply:
yes i think these should be error logs. i guess it's debatable but i saw some of the pre-existing error log calls in this file and they followed that pattern.